src/deploy/functions/containerCleaner.ts

```diff
@@ -128,6 +133,196 @@ export class ContainerRegistryCleaner {
+ * @param artifacts: a list of artifact names from GCR
+ * @param path: the relative GCR path that the artifacts live
+ */
+async function formatTagsAndInsert(
+  dockerHelper: DockerHelper,
+  artifactList: string[],
+  path: string,
+  artifacts: Set<string>
+): Promise<void> {
+  for (const artifact of artifactList) {
+    const tags = await getFlattenedTags(dockerHelper, path + `/${artifact}`);
+    const artifactEntry = tags.length === 0 ? artifact : `${artifact} - tags:${tags}`;
+    artifacts.add(artifactEntry);
+  }
+}
+
+/**
+ * Helper function to find all relative Google Cloud Registry paths for use with searching for GCF artifacts
+ * @param projectId: the current project that contains GCF artifacts
+ * @param subdomain: the current subdomain that we're searching or deleting from
+ * @param locations: an optional array that contains specific Google Regions. If ommited, will return the root path.
+ * @returns an array of relative GCR paths.
+ */
+function findGCFPaths(projectId: string, subdomain: string, locations?: string[]): string[] {
+  const paths: string[] = [];
+  if (locations) {
+    for (const location of locations) {
+      if (SUBDOMAIN_MAPPING[location] !== subdomain) {
+        continue;
+      }
+      paths.push(`${projectId}/gcf/${location}`);
+    }
+  } else {
+    paths.push(`${projectId}/gcf`);
+  }
+  return paths;
+}
+
+/**
+ * Helper function to find all the subdomains for the supplied locations
+ * @param locations: an optional array of google cloud regions, if omitted will return all {@link SUBDOMAINS}
+ * @returns a set of subdomains
+ */
+function getSubdomainsFromLocations(locations?: string[]): Set<string> {
+  if (locations === undefined) {
+    return new Set(SUBDOMAINS);
+  }
+  const subdomains = new Set<string>();
+  for (const location of locations) {
+    subdomains.add(SUBDOMAIN_MAPPING[location]);
+  }
+  return subdomains;
+}
+
+/**
+ * List all artifacts from the GCF directory in GCR.
+ * @param projectId: the current project that contains GCF artifacts
+ * @param location: the specific region to search for artifacts. If omitted, will search all locations.
+ * @param dockerHelpers: a map of {@link SUBDOMAINS} to {@link DockerHelper}. If omitted, will use the default value and create each {@link DockerHelper} internally.
+ *
+ * @throws {@link FirebaseError}
+ * Thrown if the provided location is not a valid Google Cloud region or we fail to search subdomains.
+ */
+export async function listGCFArtifacts(
+  projectId: string,
+  locations?: string[],
+  dockerHelpers: Record<string, DockerHelper> = {}
+): Promise<Set<string>> {
+  if (locations && locations.find((location) => SUBDOMAIN_MAPPING[location] === undefined)) {
+    throw new FirebaseError("Invalid region supplied.");
+  }
+
+  const artifacts = new Set<string>();
+  const failedSubdomains = new Set<string>();
+  const subdomains = getSubdomainsFromLocations(locations);
+  for (const subdomain of subdomains) {
```

Comment:
Thanks for writing this all out! I've changed the list function to use  what you wrote here. I modified the parallel work to go over the subdomains since we only need to run a `ls` at the root of the `gcf` directory. I also changed to helper to accept a subdomain instead of a location.